### PR TITLE
Add deployment intermediate states

### LINKED ISSUE #1208: Under circumstances deployment created event can be written again
### Situation
Currently this is not an real issue since the state would be same, but it seems a bit odd to have multiple created events with the same deployment.

### Scenario
* create deployment 
 * deployment created
 * broker stop
* delete snapshot
 * broker start
 * reprocessing starts
 * reprocessing create will cause push deployments again, which means will create created event on other partitions

### Possible solution
Add new command `push`, will written on `CREATE` processing to other partitions and on partition 1. On `push` processing, on the remaining partitions, the current deployment state have to be checked. If the deployment already exist, then pushed command will be rejected, otherwise  CREATED event will be written. 

For partition 1 there are two solutions:

 A) There exist an flag on the response, which indicates the successful creation. If only one response contains`true` then the created event will be written, otherwise pushed will be rejected.
B) Introduce another command, for example `creating`  which is written after all responses came back. On processing this command the deployment state is checked. If the deployment already exist, the cmd will be rejected otherwise created event will be written.

All Deployments have to be stored into rocks db to check there existence. It is also necessary for deleting the log.

### PR DESCRIPTION
**Pull request by ChrisKujawa**
* create - transforms, validates deployment and generates key
 * distribute - distributes the deployment to remaining partitions
 * creating - verifies whether the deployment was already added or not,
   if it exist already then it will be rejected otherwise the created event is written
 * created - adds the deployment to the state
 * creating and created processor runs on each partition
 * create and distribute runs only on partition 1

closes #1208

### CODE REVIEW

**Review by menski — CHANGES_REQUESTED** [review]
Please merge `DeploymentDistributeProcessor` and `DeploymentRejectedProcessor` into `DeploymentCreateProcessor`. Also merge `DeploymentCreatedProcessor` into `DeploymentCreatingProcessor`.

It would also be good if the deployment partition has the workflow as fast as possible in the workflow index. At best after `CREATE` was processed as we can then achieve that as long as the user got a positive response to the deploy request at least one partition can create instances.

Reason for that is that we in general do the state changes and responses in the command processor. Normally events are not processed by the stream processors. The reasons for this are that at the moment the command processor does the decision we already know the new state and can use it. Assume someone creates a deployment and after short time starts to create millions of instances. If these millions of create wf instance commands are written after the `CREATING` command and before the `CREATED` event all of them are rejected. Although the workflow is already known.

File: qa/integration-tests/src/test/java/io/zeebe/broker/it/data/ZeebeObjectMapperTest.java [hunk 1]
```diff
@@ -309,6 +300,43 @@ public void shouldSerializeWorkflowInstanceRecordWithoutPayload() throws Interru
...
+                  && valueType == ValueType.DEPLOYMENT
+                  && intent.equals("CREATED")
+                  && key == deploymentEvent.getKey()) {
+                latch.countDown();
+              }
+            })
+        .open();
+
+    try {
+      latch.await();
```
Comment (menski) [inline qa/integration-tests/src/test/java/io/zeebe/broker/it/data/ZeebeObjectMapperTest.java]: Please add a timeout to the `await` call and check that it returned `true`.

File: test/src/test/java/io/zeebe/test/WorkflowTest.java [hunk 2]
```diff
@@ -52,6 +55,26 @@ public void deploy() {
...
+              final String intent = r.getMetadata().getIntent();
+              if (recordType == RecordType.EVENT
+                  && valueType == ValueType.DEPLOYMENT
+                  && intent.equals("CREATED")) {
+                latch.countDown();
+              }
+            })
+        .open();
+
+    latch.await();
```
Comment (menski) [inline test/src/test/java/io/zeebe/test/WorkflowTest.java]: Please add a timeout to the `await` call and check that it returned `true`.

**Review by menski — CHANGES_REQUESTED** [review]

**Comment by menski** [comment]
I think we don't need the await hacks in the tests anymore if the workflow is added to the index after the create command was processed.

**Comment by Zelldon** [comment]
> Please merge DeploymentDistributeProcessor and DeploymentRejectedProcessor into DeploymentCreateProcessor. Also merge DeploymentCreatedProcessor into DeploymentCreatingProcessor.

Hm can do. Personally I found it clearer and it is a bit separation of concerns.

> It would also be good if the deployment partition has the workflow as fast as possible in the workflow index. At best after CREATE was processed as we can then achieve that as long as the user got a positive response to the deploy request at least one partition can create instances.

Since we removing the workflow fetching stuff in #1165 I would try to not couple the processors more then necessary. We could think about sending the response in the CREATING processing, then we can be sure that all partitions have the deployment and we can start WF instances without any problems.

> Reason for that is that we in general do the state changes and responses in the command processor. 

Disagree. The Deployment reject and create processor are not new there already exists before my changes, furthermore there exist processors for Workflow created and rejected which does the same :sweat_smile: . 

> Normally events are not processed by the stream processors. The reasons for this are that at the moment the command processor does the decision we already know the new state and can use it. 

Makes sense but is currently not the case. We can change that.

> Assume someone creates a deployment and after short time starts to create millions of instances. If these millions of create wf instance commands are written after the CREATING command and before the CREATED event all of them are rejected. Although the workflow is already known.

Yes that is the reason why I never understood your request of directly sending the response :D 
I would say that sending the response on CREATING solve this. Did you think that we then could have some timeout problems?

**Review by menski — APPROVED** [review]

**Pull request merged**